### PR TITLE
Fix crash for large HTTP request headers

### PR DESCRIPTION
**Pull request by dnadoba**
### Motivation
Fixes tests added in #658 and #659

### Modification
Gracefully handle writability change events while headers are send.

### Result
Large headers can be send. This could actually happen randomly for small headers too, especially in the HTTP/2 case. They just need to fill the buffer and trigger back pressure.

### CODE REVIEW

File: Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ClientChannelHandler.swift [hunk 1]
```diff
@@ -183,9 +183,19 @@ final class HTTP1ClientChannelHandler: ChannelDuplexHandler {
 
     private func run(_ action: HTTP1ConnectionStateMachine.Action, context: ChannelHandlerContext) {
         switch action {
-        case .sendRequestHead(let head, startBody: let startBody):
-            self.sendRequestHead(head, startBody: startBody, context: context)
+        case .sendRequestHead(let head, let sendEnd):
+            self.sendRequestHead(head, sendEnd: sendEnd, context: context)
+        case .notifyRequestHeadSendSuccessfully(let resumeRequestBodyStream, let startIdleTimer):
 
+            self.request!.requestHeadSent()
```
Comment (Lukasa) [inline Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ClientChannelHandler.swift:190]: Why is this force-unwrap safe?

File: Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ClientChannelHandler.swift [hunk 2]
```diff
@@ -183,9 +183,19 @@ final class HTTP1ClientChannelHandler: ChannelDuplexHandler {
...
+            if resumeRequestBodyStream {
+                self.request!.resumeRequestBodyStream()
```
Comment (Lukasa) [inline Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ClientChannelHandler.swift:192]: Relatedly, why is this safe? Can it end up `nil` after the preceding outcall?

File: Sources/AsyncHTTPClient/ConnectionPool/HTTP2/HTTP2ClientRequestHandler.swift [hunk 3]
```diff
@@ -140,9 +140,18 @@ final class HTTP2ClientRequestHandler: ChannelDuplexHandler {
 
     private func run(_ action: HTTPRequestStateMachine.Action, context: ChannelHandlerContext) {
         switch action {
-        case .sendRequestHead(let head, let startBody):
-            self.sendRequestHead(head, startBody: startBody, context: context)
-
+        case .sendRequestHead(let head, let sendEnd):
+            self.sendRequestHead(head, sendEnd: sendEnd, context: context)
+        case .notifyRequestHeadSendSuccessfully(let resumeRequestBodyStream, let startIdleTimer):
+            self.request!.requestHeadSent()
```
Comment (Lukasa) [inline Sources/AsyncHTTPClient/ConnectionPool/HTTP2/HTTP2ClientRequestHandler.swift:146]: Same notes here about the force unwraps.

File: Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ClientChannelHandler.swift (hunk 1 above)
Comment (dnadoba) [reply Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ClientChannelHandler.swift:191]: Added a comment

File: Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ClientChannelHandler.swift (hunk 2 above)
Comment (dnadoba) [reply Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ClientChannelHandler.swift]: Great catch! This is not safe. I added a test which crashed previously bit is fixed by chaining the implantation to unwrap safely: https://github.com/swift-server/async-http-client/pull/661/commits/18f6505fd957e3ac2e8aeca148faa95f05987b38

File: Sources/AsyncHTTPClient/ConnectionPool/HTTP2/HTTP2ClientRequestHandler.swift (hunk 3 above)
Comment (dnadoba) [reply Sources/AsyncHTTPClient/ConnectionPool/HTTP2/HTTP2ClientRequestHandler.swift:148]: Added a comment

**Comment by Lukasa** [comment]
@swift-server-bot test this please

**Review by Lukasa — APPROVED** [review]

**Pull request merged**